File: hash_tree.py
import apriori_extragen,sys

class Node:
    def __init__(self,parent, state,depth):              # initialize
                  #node with parent and state (0 = leaf, 1 = interior)
        self.parent = parent
        self.children = {}
        self.state = state
        self.item_lists = []
        self.item_sets = []
        self.trans_list = []
        self.depth = depth

    def get_parent(self):
        return self.parent

    def get_state(self):
        return self.state

    def get_children(self):
        return self.children

    def get_num_children(self):
        return len(self.children)

    def add_child(self,depth,node):
        assert(self.state == 1)

        item_list = node.item_lists[0]
        self.children[item_list[depth]] = node

    def remove_child(self,node):
        del self.children[node]

    def add_transaction(self,transaction):
        self.item_lists.append(list(sorted(transaction.item_set)))
        self.item_sets.append(transaction.item_set)
        self.trans_list.append(transaction)


    def get_num_lists(self):
        return len(self.item_lists)

    # update all three lists to reflect the change of one item set
    def append_itemset(self,item_set,trans):
        self.item_sets.append(item_set)
        self.item_lists.append(sorted(list(item_set)))
        self.trans_list.append(trans)

    # make node into an interior node
    def _make_interior(self):
        debug("leaf-> into interior")
        self.state = 1
        self.item_lists = []
        self.item_sets = []
        self.trans_list = []


    def expand_node(self,threshold):
        if len(self.item_lists) <= threshold:
            return 0;

        else:
            debug("____new call to expand_node______")
            debug("self.item_lists: " + str(self.item_lists))
            for index,itemset in enumerate(self.item_lists):
                debug("self.depth: " + str(self.depth))
                debug("itemset: " + str(itemset))
                key = itemset[self.depth]
                child = self.children.setdefault(key,\
                        Node(self,0,self.depth +1))
                child.add_transaction(self.trans_list[index])
            self._make_interior()

            [child.expand_node(threshold) for child in self.children.values()]
            return 1;




class Hash_tree:
    def __init__(self,threshold):
        self.threshold = threshold
        self.root = None
        self.cand_list_final = []
        self.L_k_set = []

    def reinitialize(self):
        self.root = None
        self.cand_list_final = []

    def add_trans(self, trans):
        debug("trans to be added: " +  str(trans.item_set))
        if self.root == None:
            self.root = Node(None,0,0)
            self.root.add_transaction(trans)
            debug(str(trans.item_set) + " put at depth zero ")

        else:
            self.add_trans_recursive(self.root,\
                    sorted(list(trans.item_set)),trans,0)

    def add_trans_recursive(self, node, list1,trans,depth):
        if node.get_state() == 1:       # if its an interior node:
            # hash on depth, if found call funct again
            debug("looking at interior")
            if list1[depth] in node.children.keys():
                self.add_trans_recursive(node.children[list1[depth]]\
                        ,list1,trans,depth + 1)
            else:          # if not found, make a new Node
                temp = Node(node,0,depth+1)
                temp.add_transaction(trans)
                debug(str(trans.item_set) + " put at depth: " +\
                        str(depth))
                node.add_child(depth,temp)
        else:               # if its a leaf node:
            debug("looking at leaf at depth: " + str(depth))

            node.add_transaction(trans)   # add the list to the node, and
                                            #then check threshold
            assert(len(node.trans_list[0].item_set)==
                    len(trans.item_set),node.trans_list)
            node.expand_node(self.threshold)




    def check_threshold(self,node): # return true if greater than
                                    # threshold
        return node.get_num_lists() > self.threshold

    def print_all_tree(self,node):

        debug("_________________")
        if node.get_state() == 1:
            debug1("in an interior node")
            for i in node.children.values():
                self.print_all_tree(i)
            debug1("out of interior node")
        else:
            debug1("Here's a leaf node item list: "+
                    str(node.item_lists))
            debug1("Here's item set: "+ str(node.item_sets))

    def prints(self):
        if self.root != None:
            self.print_all_tree(self.root)

    def subset(self,t):
        debug1("dealing with this t: " + str(t))
        if self.root != None:
            t_list = sorted(list(t.item_set))
            self.subset_recursive(self.root,0,t_list,t)
            debug1("DONE WITH SUBSET FUNCTION")


    def subset_recursive(self,node,num_to_hash,t_list,t):
        if node.get_parent() == None:      # if root, hash on every item
                                            # in t
            if node.get_state() == 1:
                for item in t_list:
                    if item in node.children.keys():
                        child = node.children[item]
                        debug1("@ root: " + str(item))
                        self.subset_recursive(child,num_to_hash+1,t_list,t)
            else:
                debug1("t.item_set: " + str(t.item_set))
                for index,item_set in enumerate(node.item_sets):
                    debug1("item_set: " + str(item_set))

                    if item_set.issubset(t.item_set):
                        debug1(str(node.trans_list[index])+ " appended")
                        self.cand_list_final.append(node.trans_list[index])



        else:                   # if not root, check interior or leaf
            if node.get_state() == 0: # if leaf, find ones in T
                debug1("t.item_set: " + str(t.item_set))
                for index,item_set in enumerate(node.item_sets):
                    debug1("item_set: " + str(item_set))

                    if item_set.issubset(t.item_set):
                        debug1(str(node.trans_list[index])+ " appended")
                        self.cand_list_final.append(node.trans_list[index])
            else:
                #if interior node, hash on items after
                debug1("in interior node")
                for index,iter in enumerate(t_list[num_to_hash:\
                        len(t_list)]):
                    debug1("iter: " + str(iter))
                    if iter in node.children.keys():
                        debug1("iter: " + str(iter) + "; num_to_hash: "\
                                + str(num_to_hash))
                        self.subset_recursive(node.children[iter],\
                            index+num_to_hash+1, t_list, t)
                        debug1("out of interior node")

    def check_minsup(self,node,minsup):
        if node != None:
            if node.get_state() == 1:
                for i in node.children.values():
                    self.check_minsup(i,minsup)
            else:
                assert(len(node.trans_list) == \
                        len(node.item_lists))
                for trans in node.trans_list:
                    debug("c in cand_trans_list: " + str(trans))
                    if trans.count >= minsup:
                        self.L_k_set.append(trans)


def debug(data):
    if False:
        #sys.stderr.write(str(data))
       # sys.stderr.write('\n')

        print data

def debug1(data):
    if False:
        print data




if __name__ == '__main__':
    #a=apriori.Transaction(set([1,3,2]))
    #b=apriori.Transaction(set([1,3,5]))
    #c=apriori.Transaction(set([1,3,4]))
    #d=apriori.Transaction(set([1,2,5]))
    #e=apriori.Transaction(set([1,2,4]))
    #f=apriori.Transaction(set([1,5,4]))
    #g=apriori.Transaction(set([3,2,5]))
    #h=apriori.Transaction(set([3,5,4]))
    #i=apriori.Transaction(set([3,2,4]))
    #j=apriori.Transaction(set([2,5,4]))
    ht = Hash_tree(3)
    #ht.add_trans(a)
    #ht.add_trans(b)
    #ht.add_trans(c)
    #ht.add_trans(d)
    #ht.add_trans(e)
    #ht.add_trans(f)
    #ht.add_trans(g)
    #ht.add_trans(h)
    #ht.add_trans(i)
    #ht.add_trans(j)
    #ht.add_trans(apriori.Transaction(set([1,2,3])))
    #ht.add_trans(apriori.Transaction(set([2,1,4])))
    #ht.add_trans(apriori.Transaction(set([1,2,5])))
    #ht.add_trans(apriori.Transaction(set([1,2,6])))
    #ht.add_trans(apriori.Transaction(set([2,1,5])))
    #ht.add_trans(apriori.Transaction(set([1,3,6])))
    #ht.add_trans(apriori.Transaction(set([3,2,6])))
    #ht.add_trans(apriori.Transaction(set([1,6,5])))
    #ht.add_trans(apriori.Transaction(set([1,2,4])))
    #ht.add_trans(apriori.Transaction(set([1,2,5])))
    #ht.add_trans(apriori.Transaction(set([1,2,6])))
    #ht.add_trans(apriori.Transaction(set([3,1,7])))
    #ht.add_trans(apriori.Transaction(set([1,3,2,5,4])))
    #ht.add_trans(apriori.Transaction(set([3,2,5,4,3])))
    #ht.add_trans(apriori.Transaction(set([2,5,4,3,1])))
    #ht.add_trans(apriori.Transaction(set([5,4,3,1,2])))
    #ht.add_trans(apriori.Transaction(set([4,3,1,2])))
    #ht.add_trans(apriori.Transaction(set([3,1,2,1])))
    #ht.add_trans(apriori.Transaction(set([1,2,3])))
    #ht.add_trans(apriori.Transaction(set([2,3,1])))
    #ht.add_trans(apriori.Transaction(set([2,3,1])))
    #ht.add_trans(apriori.Transaction(set([3,1,2,4])))
    #ht.add_trans(apriori.Transaction(set([1,2,3,4,5])))
    #ht.add_trans(apriori.Transaction(set([2,3,4,5,6])))
    #ht.add_trans(apriori.Transaction(set([3,4,5,6,7])))
    #ht.add_trans(apriori.Transaction(set([4,5,6,7,8])))
    #ht.add_trans(apriori.Transaction(set([5,6,7,8,2])))
    #ht.add_trans(apriori.Transaction(set([6,7,8,2,1])))
    #ht.add_trans(apriori.Transaction(set([7,8,2,1])))
    #ht.add_trans(apriori.Transaction(set([8,2,1,4])))
    #ht.add_trans(apriori.Transaction(set([2,1,4])))
    #ht.add_trans(apriori.Transaction(set([1,4,2])))
    #ht.add_trans(apriori.Transaction(set([1,4,2])))
    #ht.add_trans(apriori.Transaction(set([4,2,3])))
    #ht.add_trans(apriori.Transaction(set([2,3,4])))
    #ht.add_trans(apriori.Transaction(set([2,3,4,5])))
    #ht.add_trans(apriori.Transaction(set([2,3,4,5,8])))
    #ht.add_trans(apriori.Transaction(set([3,4,5,8,7])))
    ht.add_trans(apriori.Transaction(set([2, 5, 6])))
    ht.add_trans(apriori.Transaction(set([2, 4, 7])))
    ht.add_trans(apriori.Transaction(set([9, 3, 2])))
    ht.add_trans(apriori.Transaction(set([9, 8, 7])))
    ht.add_trans(apriori.Transaction(set([1, 4, 7])))
    ht.add_trans(apriori.Transaction(set([9, 3, 6])))
    ht.add_trans(apriori.Transaction(set([1, 3, 5])))
    ht.add_trans(apriori.Transaction(set([9, 2, 4])))
    ht.add_trans(apriori.Transaction(set([1, 8, 4])))
    ht.add_trans(apriori.Transaction(set([9, 8, 2])))
    ht.add_trans(apriori.Transaction(set([2, 5, 4])))
    ht.add_trans(apriori.Transaction(set([8, 2, 5])))
    ht.add_trans(apriori.Transaction(set([2, 7, 6])))
    ht.add_trans(apriori.Transaction(set([8, 5, 6])))
    ht.add_trans(apriori.Transaction(set([1, 8, 3])))
    ht.add_trans(apriori.Transaction(set([3, 7, 6])))
    ht.add_trans(apriori.Transaction(set([1, 2, 5])))
    ht.add_trans(apriori.Transaction(set([9, 4, 6])))
    ht.add_trans(apriori.Transaction(set([1, 2, 4])))
    ht.add_trans(apriori.Transaction(set([3, 5, 4])))
    ht.add_trans(apriori.Transaction(set([3, 2, 6])))
    ht.add_trans(apriori.Transaction(set([2, 5, 7])))
    ht.add_trans(apriori.Transaction(set([8, 4, 6])))
    ht.add_trans(apriori.Transaction(set([1, 3, 7])))
    ht.add_trans(apriori.Transaction(set([1, 9, 7])))
    ht.add_trans(apriori.Transaction(set([8, 2, 6])))
    ht.add_trans(apriori.Transaction(set([1, 3, 2])))
    ht.add_trans(apriori.Transaction(set([3, 5, 7])))
    ht.add_trans(apriori.Transaction(set([8, 3, 4])))
    ht.add_trans(apriori.Transaction(set([9, 2, 7])))
    ht.add_trans(apriori.Transaction(set([1, 8, 2])))
    ht.add_trans(apriori.Transaction(set([1, 9, 3])))
    ht.add_trans(apriori.Transaction(set([1, 9, 8])))
    ht.add_trans(apriori.Transaction(set([8, 3, 7])))
    ht.add_trans(apriori.Transaction(set([9, 7, 6])))
    ht.add_trans(apriori.Transaction(set([8, 3, 6])))
    ht.add_trans(apriori.Transaction(set([1, 5, 4])))
    ht.add_trans(apriori.Transaction(set([1, 8, 6])))
    ht.add_trans(apriori.Transaction(set([5, 4, 6])))
    ht.add_trans(apriori.Transaction(set([1, 9, 6])))
    ht.add_trans(apriori.Transaction(set([8, 2, 7])))
    ht.add_trans(apriori.Transaction(set([3, 2, 7])))
    ht.add_trans(apriori.Transaction(set([8, 5, 4])))
    ht.add_trans(apriori.Transaction(set([9, 3, 4])))
    ht.add_trans(apriori.Transaction(set([1, 8, 5])))
    ht.add_trans(apriori.Transaction(set([3, 2, 4])))
    ht.add_trans(apriori.Transaction(set([9, 5, 4])))
    ht.add_trans(apriori.Transaction(set([2, 4, 6])))
    ht.add_trans(apriori.Transaction(set([5, 7, 6])))
    ht.add_trans(apriori.Transaction(set([9, 8, 3])))
    ht.add_trans(apriori.Transaction(set([9, 3, 5])))
    ht.add_trans(apriori.Transaction(set([1, 5, 6])))
    ht.add_trans(apriori.Transaction(set([3, 5, 6])))
    ht.add_trans(apriori.Transaction(set([1, 2, 6])))
    ht.add_trans(apriori.Transaction(set([8, 2, 4])))
    ht.add_trans(apriori.Transaction(set([1, 4, 6])))
    ht.add_trans(apriori.Transaction(set([4, 7, 6])))
    ht.add_trans(apriori.Transaction(set([9, 4, 7])))
    ht.add_trans(apriori.Transaction(set([8, 7, 6])))
    ht.add_trans(apriori.Transaction(set([1, 9, 2])))
    ht.add_trans(apriori.Transaction(set([1, 2, 7])))
    ht.add_trans(apriori.Transaction(set([9, 5, 6])))
    ht.add_trans(apriori.Transaction(set([9, 8, 6])))
    ht.add_trans(apriori.Transaction(set([3, 2, 5])))
    ht.add_trans(apriori.Transaction(set([8, 3, 5])))
    ht.add_trans(apriori.Transaction(set([3, 4, 6])))
    ht.add_trans(apriori.Transaction(set([9, 8, 4])))
    ht.add_trans(apriori.Transaction(set([9, 3, 7])))
    ht.add_trans(apriori.Transaction(set([9, 8, 5])))
    ht.add_trans(apriori.Transaction(set([8, 3, 2])))
    ht.add_trans(apriori.Transaction(set([5, 4, 7])))
    ht.add_trans(apriori.Transaction(set([1, 8, 7])))
    ht.add_trans(apriori.Transaction(set([3, 4, 7])))
    ht.add_trans(apriori.Transaction(set([1, 9, 4])))
    ht.add_trans(apriori.Transaction(set([1, 3, 4])))
    ht.add_trans(apriori.Transaction(set([8, 5, 7])))
    ht.add_trans(apriori.Transaction(set([1, 7, 6])))
    ht.add_trans(apriori.Transaction(set([9, 5, 7])))
    ht.add_trans(apriori.Transaction(set([9, 2, 6])))
    ht.add_trans(apriori.Transaction(set([9, 2, 5])))
    ht.add_trans(apriori.Transaction(set([1, 3, 6])))
    ht.add_trans(apriori.Transaction(set([1, 5, 7])))
    ht.add_trans(apriori.Transaction(set([8, 4, 7])))
    ht.add_trans(apriori.Transaction(set([1, 9, 5])))


    debug1("ht.prints(): ")
    ht.prints()
    f = apriori.Transaction(set([1,2,3,4,5]))
    ht.subset(f)


    print ht.cand_list_final

File: c_apriori/Apriori_c/hash_tree.c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include "linked_list.h"
#include "hash_tree.h"
#include "hash_map.h"
#include "apriori.h"

#ifdef CMOCKERY
extern void* _test_malloc(const size_t size, const char* file, const int line);
extern void _test_free(void* const ptr, const char* file, const int line);
#define malloc(size) _test_malloc(size, __FILE__, __LINE__)
#define free(ptr) _test_free(ptr, __FILE__, __LINE__)
#endif 

/****************************************************************************/
/*							Hash Tree Node 									*/
/****************************************************************************/ 

void init_hash_tree_node(struct hash_tree_node *ht_node, struct hash_tree_node *parent, 
		enum state node_type, unsigned char depth)
{
	ht_node->parent = (void *) parent;
	ht_node->children = NULL;
	ht_node->node_type = node_type;
	ht_node->item_lists = NULL;
	ht_node->depth = depth;
}

void add_child(struct hash_tree_node *ht_node, uint32_t depth,
		struct hash_tree_node *child)
{
	insert_in_hash(ht_node->children,depth,child);
}
	
int add_transaction(struct hash_tree_node *ht_node, void *data)
{
	return add(&ht_node->item_lists,data,0);
}

int get_num_lists(struct hash_tree_node *ht_node)
{
	return get_len_list(ht_node->item_lists);
}

void make_interior(struct hash_tree_node *ht_node)
{
	ht_node->node_type = interior;
//	print_lists(ht_node->item_lists);
//	free_list_of_lists(&ht_node->item_lists);
	ht_node->item_lists = NULL;
}

int expand_node(struct hash_tree_node *ht_node, int threshold)
{
	int k;
	void *data;
	uint32_t key;
	struct node *present;
	struct hash_tree_node *child = NULL;
	if(get_num_lists(ht_node) <= threshold)
		return FALSE;
	struct node *next = NULL;
	struct node *iter = ht_node->item_lists;
	if(ht_node->children == NULL)
	{
		initialize_hash_map(&ht_node->children);
	}
	while(iter != NULL)
	{	
		data = get_data((struct node *)iter->data,ht_node->depth);
		if(data != NULL)
		{
			key = *((uint32_t *)data);
			present = get_data_from_hash(ht_node->children,key);
			if(present != NULL)
			{
				child = present->data;
			}
			else
			{
				child = malloc(sizeof(struct hash_tree_node));
				init_hash_tree_node(child,ht_node,leaf,ht_node->depth + 1);
				add_child(ht_node,key,child);	
			}
			
			add_transaction(child,iter->data);
		}
		next = iter->next;
		free(iter);
		iter = next;
	}
	
	make_interior(ht_node);
	
	for(k = 0; k < ht_node->children->size; k++)
	{
		if(ht_node->children->hash_table[k] != NULL)
		{
			expand_node((struct hash_tree_node *)ht_node
					->children->hash_table[k]->data,threshold);
		}
	}
	return TRUE;
}
	
void free_hash_tree_node(struct hash_tree_node **ht_node) 
{
	if(*ht_node != NULL)
	{
		if((*ht_node)->children != NULL)
		{
			free_hash_map(&(*ht_node)->children);
			
		}
		if((*ht_node)->item_lists != NULL)
		{
			free_list_of_lists(&(*ht_node)->item_lists);
		}
		free(*ht_node);
		*ht_node = NULL;
	}
	
}


void free_hash_tree_int_node(void *ht_node)
{
	free(((struct hash_tree_node *)(ht_node))->children);
	free(ht_node);
}


/****************************************************************************/
/*								Hash Tree 									*/
/****************************************************************************/ 


void init_hash_tree(struct hash_tree **ht, unsigned char threshold)
{
	(*ht) = malloc(sizeof(struct hash_tree));
	(*ht)->threshold = threshold;
	(*ht)->root = NULL;
	(*ht)->cand_list_final = NULL;
	(*ht)->l_k_set = NULL;
}

void reinit_hash_tree(struct hash_tree *ht)
{
	free_hash_tree_node(&ht->root);
	
	
}

void add_trans(struct hash_tree **ht,struct node *trans)
{
	if((*ht)->root == NULL)
	{
		(*ht)->root = malloc(sizeof(struct hash_tree_node));
		init_hash_tree_node((*ht)->root,NULL,leaf,0);
		add_transaction((*ht)->root,trans);
		
	}
	else
	{
		add_trans_recursive((*ht),(*ht)->root,mergesort(trans,&compare_ints),0);
	}
	
}

void add_trans_recursive(struct hash_tree *ht,struct hash_tree_node *ht_node, 
		struct node *trans, uint32_t depth)
{
	/* if its an interior node */ 
	uint32_t num;
	void *data;
	if(ht_node->node_type == interior)
	{
		data = get_data(trans,depth);
		if(data != NULL)
		{
			num = *((uint32_t *)data);
			struct node *ll = get_data_from_hash(ht_node->children,num);
			if(ll != NULL)
			{
				add_trans_recursive(ht, (struct hash_tree_node *)ll->data,trans,depth + 1);
			}
			else
			{
				struct hash_tree_node *temp = malloc(sizeof(struct hash_tree_node));
				init_hash_tree_node(temp,ht_node,leaf,depth+1);
				add_transaction(temp,trans);
				
				add_child(ht_node,num,temp);
				
			}
		}
	}
	/* if its a leaf node */ 
	else
	{	
		add_transaction(ht_node,trans);
		expand_node(ht_node,ht->threshold);
	}
}

void free_hash_tree(struct hash_tree *ht)
{
	if(ht->root != NULL)
	{
		free_hash_tree_node(&ht->root);
	}
	if(ht->cand_list_final != NULL)
		free_list_of_lists(&ht->cand_list_final);
	if(ht->l_k_set != NULL)
		free_list_of_lists(&ht->l_k_set);
	free(ht);
	
}

void print_all_tree(struct hash_tree_node *ht_node)
{
	if(ht_node != NULL)
	{
		if(ht_node->node_type == interior)
		{
			int k;
			
			printf("In an interior node\n");
			for(k = 0; k < ht_node->children->size; k++)	
			{
				if(ht_node->children->hash_table[k] != NULL)
				{
					print_all_tree((struct hash_tree_node *)ht_node->
							children->hash_table[k]->data);	
				}
			}
			printf("Out of interior node\n");
		}
		else
		{
			printf("Leaf node list: \n");
			print_lists(ht_node->item_lists);
		}
	}
}

void subset(struct hash_tree *ht,struct node **trans)
{
	if(ht->root != NULL)
	{
		*trans = mergesort(*trans,&compare_ints);
		subset_recursive(ht,ht->root,0,*trans);
	}
}

void subset_recursive(struct hash_tree *ht, struct hash_tree_node *ht_node,
		uint32_t num_to_hash, struct node *trans)
{	
	struct node *child = NULL;
	struct node *iter = NULL;
	/* if root, hash on every item in trans */ 
	if(ht_node->parent == NULL)
	{
		
		if(ht_node->node_type == interior)
		{
			iter = trans;
			
			while(iter != NULL)
			{
				if((child = get_data_from_hash(ht_node->children,*((uint32_t *)iter->data))))
				{
					subset_recursive(ht,(struct hash_tree_node *)child->data,num_to_hash+1,trans);
				}
				iter = iter->next;
			}
		}
		else
		{
			struct node *iter = ht_node->item_lists;
			while(iter != NULL)
			{
				
				if(is_subset(trans,(struct node *)iter->data))
				{	
					
					iter->count += 1;
					
				}
				
				iter = iter->next;
			}
		}
	}
	else
	{
		if(ht_node->node_type == leaf)
		{
			struct node *iter = ht_node->item_lists;
			while(iter != NULL)
			{
				if(is_subset(trans,(struct node *)iter->data))
				{	
					iter->count += 1;
				}
				
				iter = iter->next;
				
				
			}
		}
		else
		{
			unsigned char index = 0;
			iter = get_node(trans,num_to_hash);
			while(iter != NULL)
			{
				if((child = get_data_from_hash(ht_node->children,*((uint32_t *)iter->data))) != NULL)
				{
					subset_recursive(ht,(struct hash_tree_node *)child->data,index+num_to_hash+1,trans);
				}
				index++;
				iter = iter->next;
			}
		}
	}
}

void check_minsup(struct hash_tree *ht, struct hash_tree_node *ht_node,double minsup)
{
	if(ht_node != NULL)
	{
		if(ht_node->node_type == interior)
		{
			
			int iter;
			for(iter = 0; iter < ht_node->children->size; iter++)
			{
				if(ht_node->children->hash_table[iter] != NULL)
				{
					check_minsup(ht,(struct hash_tree_node *)ht_node->children
							->hash_table[iter]->data,minsup);
				}
			}
		}
		else
		{
			struct node *iter = ht_node->item_lists;
			struct node *copy = NULL;
			while(iter != NULL)
			{
				if(iter->count  >= minsup)
				{
					copy = copy_list(iter->data);
					add(&ht->l_k_set,copy,iter->count);
				}
				iter = iter->next;
			}
		}
		
	}
}



